### ISSUE TITLE
Can't we do this with just runc?

### ISSUE THREAD

**Issue by rhatdan** [tchughesiv/oci-uid-hook#5]
Why does this only work with docker?  Isn't the information you need provided by the runc config?  If we move to ocid in the future I want to make sure we could easily use this hook.

@mrunalp Shouldn't this code work for runc?

**Comment by tchughesiv** [comment]
when i was relying on docker api, it was a docker-only deal... but now, it could potentially work w/ just runc.  So long as i can get assigned uid & image ID from runc config.json

**Comment by rhatdan** [comment]
I believe you can, and then your hook could be used by any container runtime that uses runc.

**Comment by rhatdan** [comment]
Theoretically any OCI compliant runtime that supports entering the mount namespace.

**Comment by tchughesiv** [comment]
@rhatdan @mrunalp I'm not seeing where i can grab the container's image ID/spec w/ runc.  Any ideas?

**Comment by mrunalp** [comment]
The issue is that we don't have any image metadata as a first class field in the runc config.json. We have two potential options:
1. Set an environment variable container=cri-o for the hooks
2. Set image metadata as annotations in the config.json 

Wtih either of these we can use containers/image to get the image metadata from a different location.

**Comment by rhatdan** [comment]
Why do you need the image ID?  Isn't all you need the path to the ROOTFS?

**Comment by tchughesiv** [comment]
I have to determine the user that the image was configured to run as.

**Comment by tchughesiv** [comment]
So I can do a string replace on the correct passwd entry

**Comment by rhatdan** [comment]
If I run a container with uid 3267
```
docker run -ti -u 3267 fedora sh
$ id
uid=3267 gid=0(root) groups=0(root)
```
Then I go look in the config.json file generated by docker

```
 cat /run/docker/libcontainerd/10b639cf59b36f6189b38a67f91a7812e289f3f0a2e9672066c54bfea87e9a15/config.json | json_pp | less

   "process" : {
          ...
      "user" : {
         "uid" : 3267,
         "gid" : 0
      },


   "root" : {
      "path" : "/var/lib/docker/overlay2/fe9a2094f7a41daf59d204fdb35a9f8fbfaa7ca
53ad55c4e43e9063df21c9def/merged"
   },
```
This gives me the uid and the rootfs to modify the 
/var/lib/docker/overlay2/fe9a2094f7a41daf59d204fdb35a9f8fbfaa7ca
53ad55c4e43e9063df21c9def/merged/etc/passwd
file.  What other data do you need?

**Comment by tchughesiv** [comment]
i need the .Config.User metadata from the image which sets the default/expected user at container runtime.
In the case of the fedora image, that would be "", which translates to 0 uid.

Using another image as an example...
```shell
$ docker inspect -f {{.Config.User}} registry.access.redhat.com/rhscl/nginx-18-rhel7
1001

$ docker run -tiu 243637 registry.access.redhat.com/rhscl/nginx-18-rhel7 sh
$ cat /run/docker/libcontainerd/17a968bfbb33ef0b22bc83c3012c995b4de7eadd0db3c0472bfe9fe6340c3e46/config.json | json_pp | less

   "process" : {
          ...
      "user" : {
         "uid" : 243637,
         "gid" : 0
      },
```
Both fields, container and image user, are critical here. Having the image user allows the hook to affect the proper user in /etc/passwd.

**Comment by rhatdan** [comment]
Ok I get it now.  But this kind of sucks.  So right now docker does not tell runc about this USER Field, if the container creation overrode it?

**Comment by tchughesiv** [comment]
correct, based on what i've seen.

**Comment by mrunalp** [comment]
@rhatdan Yeah, well, runtime spec only cares about the path to the rootfs. It doesn't care about how it was created which makes sense from separation of concerns point of view. We can add annotations when we generate the config.json from cri-o.

**Comment by rhatdan** [comment]
What does the entry in the /etc/passwd file end up looking like?

**Comment by tchughesiv** [comment]
in the above example... like this:
```shell
default:x:243637:0:Default Application User:/opt/app-root/src:/sbin/nologin
```
vs the original:
```shell
default:x:1001:0:Default Application User:/opt/app-root/src:/sbin/nologin
```
